# Basic Visualization

## Exploring Datasets with Pandas

In [28]:
import numpy as np
import pandas as pd

In [29]:
print(pd.__version__)

1.4.3


import set data Imigrasi Kanada dengan menggunakan metode pandas read_excel().

Modul untul membaca file excel : model xlrd

In [30]:
#pip install xlrd

In [31]:
#pip install openpyxl

In [32]:
df_can = pd.read_excel('https://github.com/ardhiraka/PFDS_sources/blob/master/Canada.xlsx?raw=true',
                        sheet_name='Canada by Citizenship',
                        skiprows=range(20),
                        skipfooter=2) # engine='openpyxl'

print('Data read into a pandas dataframe')


Data read into a pandas dataframe


In [33]:
df_can.head() # melihat data teratas
# tips : Banyak rows yang ingin dilihat dpat dispesifikan ddengan cara : df_can.head(10)

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

In [34]:
# melihat data terbawah
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

In [35]:
#df_can.dropna(inplace=true)

df_can.dropna(axis=1, how='all', inplace=True)

In [36]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

In [37]:
df_can.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    int64 
 10  1981      195 non-null    int64 
 11  1982      195 non-null    int64 
 12  1983      195 non-null    int64 
 13  1984      195 non-null    int64 
 14  1985      195 non-null    int64 
 15  1986      195 non-null    int64 
 16  1987      195 non-null    int64 
 17  1988      195 non-null    int64 
 18  1989      195 non-null    int64 
 19  1990      195 non-null    int64 
 20  1991      195 non-null    int64 
 21  1992      195 no

In [38]:
# convert 1980-2013 menjadi bertipe integer
# semua baris dengan kolom 1980-2013
df_can.loc[:, 1980:2013] = df_can.loc[:, 1980:2013].astype('Int64')

In [39]:
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    Int64 
 10  1981      195 non-null    Int64 
 11  1982      195 non-null    Int64 
 12  1983      195 non-null    Int64 
 13  1984      195 non-null    Int64 
 14  1985      195 non-null    Int64 
 15  1986      195 non-null    Int64 
 16  1987      195 non-null    Int64 
 17  1988      195 non-null    Int64 
 18  1989      195 non-null    Int64 
 19  1990      195 non-null    Int64 
 20  1991      195 non-null    Int64 
 21  1992      195 no

In [40]:
# cek kolom value
df_can.columns.values

array(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013], dtype=object)

In [41]:
# cek index value
df_can.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

In [42]:
# size of dataframe (rows, columns)
df_can.shape

(195, 43)

#### Hapus Kolom
Bersihkan dataset dengan menghapus beberapa kolom yang tidak perlu. Hal ini bisa dilakukan dengan metode pandas drop()

In [43]:
# kolom AREA, REG, DEV, Type, Coverage tidak diperlukan sehingga kolom tersebut dapat dihapus/dihilangkan
df_can.drop(['AREA', 'REG', 'DEV', 'Type', 'Coverage'], axis=1, inplace=True) # axis=1 : dua dimensi
df_can.head(2) 

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

#### Ubah Nama Kolom
Mengganti nama kolom dengan metode rename() dan memasukkan dictionary nama lama dan nama baru

In [44]:
# untuk baris yang ada null dihapus
df_can.dropna(axis=0, inplace=True)
df_can.tail(2)

OdName AreaName         RegName             DevName  1980  1981  1982  \
193    Zambia   Africa  Eastern Africa  Developing regions    11    17    11   
194  Zimbabwe   Africa  Eastern Africa  Developing regions    72   114   102   

     1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  2011  \
193     7    16     9  ...    56    91    77    71    64    60   102    69   
194    44    32    29  ...  1450   615   454   663   611   508   494   434   

     2012  2013  
193    46    59  
194   437   407  

[2 rows x 38 columns]

In [45]:
df_can.drop(df_can.tail(2).index, inplace=True) # drop last n rows

In [46]:
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 38 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   OdName    193 non-null    object
 1   AreaName  193 non-null    object
 2   RegName   193 non-null    object
 3   DevName   193 non-null    object
 4   1980      193 non-null    Int64 
 5   1981      193 non-null    Int64 
 6   1982      193 non-null    Int64 
 7   1983      193 non-null    Int64 
 8   1984      193 non-null    Int64 
 9   1985      193 non-null    Int64 
 10  1986      193 non-null    Int64 
 11  1987      193 non-null    Int64 
 12  1988      193 non-null    Int64 
 13  1989      193 non-null    Int64 
 14  1990      193 non-null    Int64 
 15  1991      193 non-null    Int64 
 16  1992      193 non-null    Int64 
 17  1993      193 non-null    Int64 
 18  1994      193 non-null    Int64 
 19  1995      193 non-null    Int64 
 20  1996      193 non-null    Int64 
 21  1997      193 no

In [47]:
# mengganti nama kolom dengan replace(), nama_lama : nama_baru
df_can.rename(columns={'OdName' : 'Country', 'AreaName' : 'Continent', 'RegName' : 'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

#### Menambahkan dan Menjumlahkan Kolom
Meenambahkan kolom 'Total' yang merangkum total imigran menurut negara selama periode 1980-20113

In [48]:
df_can['Total'] = df_can.sum(axis=1, numeric_only=True) # horizontal/columns, axis=0 : vertical/row
df_can.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   
3     0     0     0     0  ...     0     1     0     0     0     0     0   
4     0     0     0     0  ...     0     1     1     0     0     0     0   

   2012  2013  Total  
0  2635  2004  58639  
1   620   603  15699  
2  3774  4331  69439  
3     0     0      6  
4     1     1     15  

[5 rows x 39 columns]

In [49]:
# tipe data dionvert menjadi integer
df_can['Total'] = df_can['Total'].astype('Int64')
df_can.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   
3     0     0     0     0  ...     0     1     0     0     0     0     0   
4     0     0     0     0  ...     0     1     1     0     0     0     0   

   2012  2013  Total  
0  2635  2004  58639  
1   620   603  15699  
2  3774  4331  69439  
3     0     0      6  
4     1     1     15  

[5 rows x 39 columns]

In [50]:
# grouping berdasarkan region kemudian dijumlahkan
print(df_can.groupby('Region').sum())

# atau bisa juga grouping berdasarkan kolom
#df_can.groupby(['col_1', 'col_2'], as_index=False).mean()
#print(df_can.groupby('Continent').sum())

                            1980   1981   1982   1983   1984   1985   1986  \
Region                                                                       
Australia and New Zealand   1304   1119    848    457    481    467    532   
Caribbean                   7045   8310   8326   6998   5553   6048   8716   
Central America              734    921   1612   3648   4087   4862   5909   
Central Asia                   0      0      0      0      0      0      0   
Eastern Africa              1388   1510   1313   1043   1139   1096   1396   
Eastern Asia                6836   8895   5481   3254   2624   2979   3416   
Eastern Europe              1467   3698   6873   5459   4433   3969   6144   
Melanesia                    632    707    816    558    391    448    362   
Micronesia                     1      0      0      1      0      0      0   
Middle Africa                 33     27     44     32     42     30     61   
Northern Africa             1100   1268   1348    936    842    

#### Melihat Banyak Object null
Object null dapat dilihat dengan metode describe()

In [51]:
df_can.isnull().sum()
# tidak ada null

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

#### Melihat Ringkasan Kolom
Melihat ringkasan singat (statistika seperti mean, std, dll.) dari setiap kolom dalam kerangka data menggunakan metode describe()

In [52]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    193.000000    193.000000    193.000000    193.000000    193.000000   
mean     513.233161    572.186528    539.678756    391.186528    380.150259   
std     1959.128429   2163.209028   1876.049134   1210.018877   1203.927805   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     12.000000   
75%      262.000000    296.000000    290.000000    176.000000    183.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   193.000000   193.000000    193.000000    193.000000    193.000000   
mean    362.383420   445.544041    697.823834    721.051813    850.658031   
std    1084.355563  1231.216234   2119.126481   2455.413714   2567.264913   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.000000      0.000000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     29.000000     41.000000   
75%     198.000000   257.000000    446.000000    410.000000    523.000000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    193.000000    193.000000    193.000000    193.000000   
mean   ...   1330.316062   1277.336788   1200.367876   1255.813472   
std    ...   4447.762742   3945.732194   3460.264748   3712.490592   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.000000     24.000000     30.000000     30.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    851.000000    847.000000    929.000000    950.000000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    193.000000    193.000000    193.000000    193.000000    193.000000   
mean    1286.010363   1431.917098   1273.010363   1325.072539   1331.974093   
std     3848.112102   4484.608314   4049.649073   4268.156074   4258.468721   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       35.000000     39.000000     37.000000     42.000000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      904.000000    933.000000    804.000000    787.000000    818.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

               Total  
count     193.000000  
mean    33154.808290  
std     92217.764803  
min         1.000000  
25%       952.000000  
50%      5018.000000  
75%     22624.000000  
max    691904.000000  

[8 rows x 35 columns]

### Pandas : Indexing and Selection (Slicing)
Select Column There are two ways to filter on a column name : 

Note : Metode 1 cepat dan mudah tetapi  hanya berfungsi jika nama kolom TIDAK memiliki spasi atau karakter khusus.

Contoh :

Metode 1 :

df.column_name                  # return/mengembalikan series


Metode 2 :

df['column']                    # return series

df[['column 1], 'column 2]      # return dataframe


In [53]:
df_can.Country  # return a series

0                             Afghanistan
1                                 Albania
2                                 Algeria
3                          American Samoa
4                                 Andorra
                      ...                
188                               Vanuatu
189    Venezuela (Bolivarian Republic of)
190                              Viet Nam
191                        Western Sahara
192                                 Yemen
Name: Country, Length: 193, dtype: object

In [54]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # return dataframe
# 'Country : string ; years (1980, 1981, ...) : integer

Country  1980  1981  1982  1983  1984  1985
0                           Afghanistan    16    39    39    47    71   340
1                               Albania     1     0     0     0     0     0
2                               Algeria    80    67    71    69    63    44
3                        American Samoa     0     1     0     0     0     0
4                               Andorra     0     0     0     0     0     0
..                                  ...   ...   ...   ...   ...   ...   ...
188                             Vanuatu     0     0     0     0     0     0
189  Venezuela (Bolivarian Republic of)   103   117   174   124   142   165
190                            Viet Nam  1191  1829  2162  3404  7583  5907
191                      Western Sahara     0     0     0     0     0     0
192                               Yemen     1     2     1     6     0    18

[193 rows x 7 columns]

#### Select Row
Ada 2 cara utama untuk memilih baris :

df.loc[label]
#filter berdasarkan label dari index/column df.iloc[index]
#filter berdasarkan posisi dari index/column

Perhatikan bahwa indeks default dari dataset adalah rentang numerik dari 0 hingga 194. Hal ini membuatnya sangat sulit untuk melakukan kueri berdasarkan negara tertentu

Misalnya untuk mencari data di Jepang maka perlu mengetahui nilai indeks yang sesuai

Hal ini dapat diperbaiki dengan menyetel kolom 'Country' sebagai indeks menggunakan metode set_index()

In [56]:
df_can.index

RangeIndex(start=0, stop=193, step=1)

In [57]:
df_can.set_index('Country', inplace=True)
# Kebalikan dari set adalah reset
# Dengan demikian, untuk mereset index dapat menggunakan df_can.reset_index()

In [58]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013  Total  
Country                               
Afghanistan  2203  2635  2004  58639  
Albania       539   620   603  15699  
Algeria      4325  3774  4331  69439  

[3 rows x 38 columns]

#### Remove the name of the index (optional)

In [59]:
df_can.index.name = None

In [60]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013  Total  
Afghanistan  2203  2635  2004  58639  
Albania       539   620   603  15699  
Algeria      4325  3774  4331  69439  

[3 rows x 38 columns]

Contoh :

Melihat jumlah imigran dari Jepang (baris 87) untuk skenario sebagai berikut:

1. Data baris lengkap (semua kolom)
2. Untuk tahun 2013
3. Untuk tahun 1980 - 1985

In [61]:
# 1. The full row data (all columns), index : Japan
japan = df_can.loc['Japan']
print(japan)

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [63]:
# alternate methods 
# menggunakan posisi index
print(df_can.iloc[87])

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [65]:
print(df_can[df_can.index == 'Japan'].T.squeeze())
# T = transform

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [66]:
# 2. For year 2013
print(df_can.loc['Japan', 2013])

# alternate method
print(df_can.iloc[87, 36]) # year 2013 is the last column with a positional index of 36

982
982


In [68]:
# 3. For year 1980 to 1985
print(df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1985]])
print(df_can.iloc[87, [3, 4, 5, 6, 7, 8]])

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object
1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object


#### Mengubah Tipe Data dari Nama Kolom
Nama kolom berupa bilangan bulat seperti tahun (1980 1981, 1982, ...) dapat menimbulkan kebingungan/kerancuan.

contoh : misalkan kita ingin mengacu pada tahun 2013, tetapi orang dapat bingung dan mengira bahwa 2013 adalah indeks possisi 2013

Untuk menghindari ambiguitas tersebut, mari ubah naa kolom menjadi string '1980' - '2013'

In [69]:
df_can.columns

Index(['Continent',    'Region',   'DevName',        1980,        1981,
              1982,        1983,        1984,        1985,        1986,
              1987,        1988,        1989,        1990,        1991,
              1992,        1993,        1994,        1995,        1996,
              1997,        1998,        1999,        2000,        2001,
              2002,        2003,        2004,        2005,        2006,
              2007,        2008,        2009,        2010,        2011,
              2012,        2013,     'Total'],
      dtype='object')

In [71]:
# Karena kita  mengonversi tahun menjadi string
#  mari kita deklarasikan variabel yang akan memungkinnkan kita untuk memanggil rentang tahun dengan mudah
#  useful for plottiing later on
years = list(map(str, range(1980, 2014))) 
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

In [73]:
# convert to string
df_can.columns = list(map(str, df_can.columns))

In [74]:
df_can.columns

Index(['Continent', 'Region', 'DevName', '1980', '1981', '1982', '1983',
       '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992',
       '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001',
       '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010',
       '2011', '2012', '2013', 'Total'],
      dtype='object')

#### Filtering based on a criteria
Untuk memfilter dataframe berdasarkan suatu kondisi, kita cukup meneruskan kondisi tersebut sebagai vektor boolean

Misalnya memfilter kerangka data untuk menampilkan data di negara-negara Asia (AreaName = Asia)

In [76]:
# 1. Create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan                            True
Albania                               False
Algeria                               False
American Samoa                        False
Andorra                               False
                                      ...  
Vanuatu                               False
Venezuela (Bolivarian Republic of)    False
Viet Nam                               True
Western Sahara                        False
Yemen                                  True
Name: Continent, Length: 193, dtype: bool


In [77]:
# 2. Ambil data yang bernilai True, Pass this condition into the dataframe
df_can[condition].head()

Continent         Region             DevName  1980  1981  1982  \
Afghanistan      Asia  Southern Asia  Developing regions    16    39    39   
Armenia          Asia   Western Asia  Developing regions     0     0     0   
Azerbaijan       Asia   Western Asia  Developing regions     0     0     0   
Bahrain          Asia   Western Asia  Developing regions     0     2     1   
Bangladesh       Asia  Southern Asia  Developing regions    83    84    86   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Armenia         0     0     0     0  ...   224   218   198   205   267   252   
Azerbaijan      0     0     0     0  ...   359   236   203   125   165   209   
Bahrain         1     1     3     0  ...    12    12    22     9    35    28   
Bangladesh     81    98    92   486  ...  4171  4014  2897  2939  2104  4721   

             2011  2012  2013  Total  
Afghanistan  2203  2635  2004  58639  
Armenia       236   258   207   3310  
Azerbaijan    138   161    57   2649  
Bahrain        21    39    32    475  
Bangladesh   2694  2640  3789  65568  

[5 rows x 38 columns]

In [78]:
# we can pass multiple criteria in the same linne
# Contoh : filter AreaName = Asia dan RegName = Southern Asia

df_can[(df_can['Continent'] == 'Asia') & (df_can['Region'] == 'Southern Asia')]

# note : Ketika ingin menggunakan operator 'and' dan 'or', pandas mengharuskan untuk menggunakan '&' dan '|'
# jangan lupa untuk mengapit kondisi dengan tanda kurung

Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013   Total  
Afghanistan                  2004   58639  
Bangladesh                   3789   65568  
Bhutan                        487    5876  
India                       33087  691904  
Iran (Islamic Republic of)  11291  175923  
Maldives                        1      30  
Nepal                        1308   10222  
Pakistan                    12603  241600  
Sri Lanka                    2394  148358  

[9 rows x 38 columns]

In [82]:
# cara penulisan lain

df_can[condition & (df_can['Region'] == 'Southern Asia')]


Continent         Region             DevName  1980  \
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013   Total  
Afghanistan                  2004   58639  
Bangladesh                   3789   65568  
Bhutan                        487    5876  
India                       33087  691904  
Iran (Islamic Republic of)  11291  175923  
Maldives                        1      30  
Nepal                        1308   10222  
Pakistan                    12603  241600  
Sri Lanka                    2394  148358  

[9 rows x 38 columns]

In [83]:
print('data dimensions : ', df_can.shape)

data dimensions :  (193, 38)


## Visualizing Data using Matplotlib

Visualisasi data dilihat oleh berbagai bidang ilmu sebagai salah satu alat komunikasi modern

Tujuan utama :
Mengkomunikasikan informasi secara jelas dan efisien kepada penggunanya melali grafik informasi yang dipilih seperti tabel atau grafik

Visualisasi yang efektif mampu membantu memudahkan penggunanya dalam menganalisa dan memahami data dan bukti yang kompleks melalui objek visual (contohnya : titik, garis, batang, dll.)


Ada 2 jenis visualisasi data yang paling populer yaitu dashboard dan infografis. Kedua visualisasi ini mengguanakan kombinasi grafik, teks, dan gambar untuk mengkomunikasikan pesan data

Ketika melakukan explanatory analysis ada beberapa hal yang harus diperhatikan, yaitu :
1. Siapa audiencenya? Agar iisuaisasi yang ditampilkan menjadi relevan      (WHO?)
2. Apa yang ingin kita beritahu/informasikan kepada audience?                 (WHAT?)
3. Bagaimana kita menjelaskan dalam penggunaan data                  (HOW?)